op-batcher/batcher/espresso.go

```diff
@@ -134,45 +119,266 @@ func (l *BatchSubmitter) submitToEspresso(txdata txData, sig, batcherSignature [
 	ticker := time.NewTicker(transactionFetchInterval)
 	defer ticker.Stop()
 
-	var txQueryData espressoCommon.TransactionQueryData
 Loop:
 	for {
 		select {
 		case <-ticker.C:
-			txQueryData, err = l.Espresso.FetchTransactionByHash(l.shutdownCtx, txHash)
+			_, err = l.Espresso.FetchTransactionByHash(l.shutdownCtx, txHash)
 			if err == nil {
 				break Loop
 			}
-			l.Log.Warn("Retry fetching transaction by hash", "txHash", txHash, "error", err)
 		case <-timer.C:
 			l.Log.Error("Failed to fetch transaction by hash after multiple attempts", "txHash", txHash)
-			l.recordFailedDARequest(txdata.ID(), err)
-			return nil, fmt.Errorf("failed to fetch transaction by hash: %w", err)
+			return fmt.Errorf("failed to fetch transaction by hash: %w", err)
+		case <-ctx.Done():
+			l.Log.Info("Cancelling transaction publishing", "txHash", txHash)
+			break Loop
 		}
 	}
 
-	rawHeader, err := l.Espresso.FetchRawHeaderByHeight(l.shutdownCtx, txQueryData.BlockHeight)
+	return nil
+}
+
+func (l *BatchSubmitter) queueBlockToEspreso(ctx context.Context, block *types.Block) error {
+	batch, _, err := derive.BlockToSingularBatch(l.RollupConfig, block)
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("failed to derive batch from block: %w", err)
+	}
+
+	espressoBatch := espresso.EspressoBatch{
+		Header:   *block.Header(),
+		BlockNum: block.NumberU64(),
+		Batch:    *batch,
 	}
 
-	var header espressoCommon.HeaderImpl
-	err = json.Unmarshal(rawHeader, &header)
+	transaction, err := espressoBatch.ToEspressoTransaction(ctx, l.RollupConfig.L2ChainID.Uint64(), l.ChainSigner, l.BatcherAddress)
 	if err != nil {
-		return nil, fmt.Errorf("could not unmarshal header from bytes")
+		return fmt.Errorf("failed to create Espresso transaction from a batch: %w", err)
 	}
 
-	height := header.Header.GetBlockHeight()
+	go func() {
+		for {
+			err := l.tryPublishBatchToEspresso(ctx, *transaction)
+			if err == nil {
+				l.Log.Info(fmt.Sprintf("Published block %s to Espresso", eth.ToBlockID(block)))
+				break
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Periodically refreshes the sync status and polls Espresso streamer for new batches
+func (l *BatchSubmitter) espressoBatchLoadingLoop(ctx context.Context, wg *sync.WaitGroup, publishSignal chan struct{}) {
+	l.Log.Info("Starting EspressoBatchLoadingLoop")
+
+	defer wg.Done()
+	ticker := time.NewTicker(l.Config.PollInterval)
+	defer ticker.Stop()
+
+	streamer := espresso.EspressoStreamer{
+		BatcherAddress: l.BatcherAddress,
+		Namespace:      l.RollupConfig.L2ChainID.Uint64(),
+
+		L1Client:            l.L1Client,
+		EspressoClient:      l.Espresso,
+		EspressoLightClient: l.EspressoLightClient,
+		Log:                 l.Log,
+
+		BatchPos: 1,
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			newSyncStatus, err := l.getSyncStatus(ctx)
+			if err != nil {
+				l.Log.Error("failed to refresh sync status", "err", err)
+				continue
+			}
+
+			func() {
```

Comment:
Why use what looks like an anomymous function?

Reply:
So that I can ensure I treat the mutex correctly with `defer`s. Pulled it out into a named function for clarity.